frontend/ui/lib/router/router.dart

```diff
@@ -0,0 +1,45 @@
+import 'package:flutter/material.dart';
+import 'package:tripmeout/pages/create_trip_page.dart';
+import 'package:tripmeout/pages/trip_list_page.dart';
+import 'package:tripmeout/pages/trip_view_page.dart';
+import 'package:tripmeout/services/trip_service.dart';
+
+class Router {
+  static final String tripListRoute = '/trips';
+  static final String createTripRoute = '/trips/new';
+  static final RegExp tripViewRouteRegExp = RegExp(r'/trips/(?<tripId>[^/]+)$');
```

Comment:
We'll need a way for the application to generate routes that match this regex. I'm guessing the best way to do that here is with a static method, e.g.:

```dart
static String tripViewRoute(String tripId) {
  ...
}
```

Reply:
Yeah, I talked a bit with @chuckluv about this already. We probably want to move some of this stuff into static methods on our pages at some point, but I'm fine waiting to do that.

We should also note that these type of pages will generally have two sets of params, route params and service params. For `TripViewPage` I would expect the route param to be the trip id and the service params to be `TripService` and `PlaceVisitService`.

I would expect all of our pages to have two mirror methods that have a signature roughly like:
```dart
class SomePage {
  static final RegExp somePageRegExp = RegExp(r'/some/(?<someRouteParam>[^/]+)$');
  final String someRouteParam;
  final SomeService serviceParam;

  SomePage(this.someRouteParam, {this.serviceParam});

  static String routeFor(String someRouteParam) {
    return '/some/$someRouteParam'
  }

  static bool servesRoute(String route) {
    return somePageRegExp.firstMatch(route);
  }
  
  /// Create SomePage from a route.
  static SomePage fromRoute(String route, {SomeService serviceParam}) {
    RegExpMatch match = somePageRegExp.firstMatch(route).;
    if (match == null) {
      return null;
    }
    String someRouteParam = match.namedGroup('someRouteParam');
    return SomePage(someRouteParam, serviceParam: serviceParam);
  }
}
```

I'm actually not sure if dart has a nice optional we could use instead of the `servesRoute` and `fromRoute` pair. Or we could just always check for nullness.

This would enable a router that does something like:
```dart
SomePage somePage = SomePage.fromRoute(settings.name, serviceParam: serviceParam);
if (somePage != null) {
  return MaterialPageRoute(
    builder: (context) => somePage,
    settings: settings,
  }
}
// handle other pages...
```

Likewise when routing to a page you can do something like this so that you don't need to hardcode values everywhere:
```dart
Navigator.pushNamed(context, SomePage.routeFor(someRouteParam)); 
```

Again I don't think these are necessary in this CL, but it might be nice to have them in a follow up.

One other thing we could do is wrap the whole `Navigator.pushNamed` call to ensure we never preform self-routes, but that's more of an optimization.

Reply:
FYI, you can actually use a [`factory` constructor](https://dart.dev/guides/language/language-tour#factory-constructors) above for `fromRoute`.